### PR TITLE
[Tutorial] Deploy Quantized Model on CUDA

### PR DESCRIPTION
**Pull request by vinx13**
This tutorial demonstrates how to import a model using Relay frontend, run quantization and calibration passes, and perform quantized inference.
ref https://github.com/apache/incubator-tvm/issues/4435

cc @tqchen @masahi @anijain2305 @ZihengJiang @tmoreau89

### CODE REVIEW

File: tutorials/frontend/deploy_quantized.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Deploy a Quantized Model on Cuda
+================================
+**Author**: `Wuwei Lin <https://github.com/vinx13>`_
+
+This article is an introductory tutorial of automatic quantization with TVM.
```
Comment (anijain2305) [inline tutorials/frontend/deploy_quantized.py]: Maybe add a link to this discuss forum - https://discuss.tvm.ai/t/quantization-story/3920 to give a high-level idea of whats automatic quantization.

File: tutorials/frontend/deploy_quantized.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        std_r=std_rgb[0],
+        std_g=std_rgb[1],
+        std_b=std_rgb[2],
+    )
+    return val_data, batch_fn
+
+
+###############################################################################
+# The calibration dataset should be a iterable object. We define the
+# calibration dataset as a generator object in Python. In this tutorials, we
```
Comment (anijain2305) [inline tutorials/frontend/deploy_quantized.py]: tutorials -> tutorial

File: tutorials/frontend/deploy_quantized.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+# weights. Two modes are supported: `power2` and `max`. Both modes find the
+# maximum value within the weight tensor first. In `power2` mode, the maximum
+# is rounded down to power of two. If the scales of both weights and outputs
+# are power of two, we can leverage bit shifting for multiplications. This make
+# it computationally more efficient. In `max` mode, the maximum is used as the
+# scale. Without rounding, `max` mode might have better accuracy in some cases.
+# When the scales are not power of two, fixed point multiplications will
+# be used.
+#
+# For outputs, we can find the scales with data-aware quantization.
```
Comment (anijain2305) [inline tutorials/frontend/deploy_quantized.py]: outputs --> intermediate feature maps

File: tutorials/frontend/deploy_quantized.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+**Author**: `Wuwei Lin <https://github.com/vinx13>`_
+
+This article is an introductory tutorial of automatic quantization with TVM.
+We will import a GluonCV pre-trained model on ImageNet to Relay, quantize the Relay model
+and then perform the inference.
+"""
+
+import tvm
+from tvm import relay
+from tvm.relay import quantize as qtz
```
Comment (anijain2305) [inline tutorials/frontend/deploy_quantized.py]: Not used

File: tutorials/frontend/deploy_quantized.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        if i * batch_size >= calibration_samples:
+            break
+        data, _ = batch_fn(batch)
+        yield {'data': data}
+
+
+###############################################################################
+# Import the model
+# ----------------
+# We use the Relay MxNet frontent to import a model from the Gluon model zoo.
```
Comment (masahi) [inline tutorials/frontend/deploy_quantized.py]: frontend

File: tutorials/frontend/deploy_quantized.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+# is rounded down to power of two. If the scales of both weights and outputs
+# are power of two, we can leverage bit shifting for multiplications. This make
+# it computationally more efficient. In `max` mode, the maximum is used as the
+# scale. Without rounding, `max` mode might have better accuracy in some cases.
+# When the scales are not power of two, fixed point multiplications will
+# be used.
+#
+# For outputs, we can find the scales with data-aware quantization.
+# Data-aware quantization takes a calibration dataset as the input argument.
+# Scales are calculated by minimizing the KL divergence of between the data
```
Comment (masahi) [inline tutorials/frontend/deploy_quantized.py]: of between -> between
I'd prefer "KL divergence between distribution of activation before and after quantization"

Comment (vinx13) [reply tutorials/frontend/deploy_quantized.py]: it is not necessarily 'activation', so i updated to 'output of each layer'

Comment (masahi) [reply tutorials/frontend/deploy_quantized.py]: By activation I do mean 'output of each layer' or intermediate feature map, not necessary the output of activation function like relu. I think this is a standard terminology in the literature (they talk about "weight 8bit, acitvation 32 bit" etc).

File: tutorials/frontend/deploy_quantized.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Deploy a Quantized Model on Cuda
+================================
+**Author**: `Wuwei Lin <https://github.com/vinx13>`_
+
+This article is an introductory tutorial of automatic quantization with TVM.
+Automatic quantization is one of the quantization mode in TVM. More details of the quantization story in TVM can be found `here <https://discuss.tvm.ai/t/quantization-story/3920>`_.
```
Comment (tmoreau89) [inline tutorials/frontend/deploy_quantized.py]: mode -> modes
details of -> details on

Comment (tmoreau89) [reply tutorials/frontend/deploy_quantized.py]: also long line

File: tutorials/frontend/deploy_quantized.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+batch_size = 1
+model_name = "resnet18_v1"
+target = 'cuda'
+ctx = tvm.context(target)
+
+###############################################################################
+# Prepare the Dataset
+# -------------------
+# We will demonstrate how to prepare the calibration dataset for quantization.
+# We first download the validate set of ImageNet and pre-process the dataset.
```
Comment (tmoreau89) [inline tutorials/frontend/deploy_quantized.py]: validation set

File: tutorials/frontend/deploy_quantized.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        mean_b=mean_rgb[2],
+        std_r=std_rgb[0],
+        std_g=std_rgb[1],
+        std_b=std_rgb[2],
+    )
+    return val_data, batch_fn
+
+
+###############################################################################
+# The calibration dataset should be a iterable object. We define the
```
Comment (tmoreau89) [inline tutorials/frontend/deploy_quantized.py]: "should be an"

File: tutorials/frontend/deploy_quantized.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+# feature map tensor of each layer.
+#
+# For weights, the scales are directly calculated based on the value of the
+# weights. Two modes are supported: `power2` and `max`. Both modes find the
+# maximum value within the weight tensor first. In `power2` mode, the maximum
+# is rounded down to power of two. If the scales of both weights and
+# intermediate feature maps are power of two, we can leverage bit shifting for
+# multiplications. This make it computationally more efficient. In `max` mode,
+# the maximum is used as the scale. Without rounding, `max` mode might have
+# better accuracy in some cases. When the scales are not power of two, fixed
```
Comment (tmoreau89) [inline tutorials/frontend/deploy_quantized.py]: powers

**Comment by masahi** [comment]
Thanks @vinx13  @anijain2305 @tmoreau89

**Pull request merged**